### PR TITLE
Add support for 302 redirect

### PR DESCRIPTION
**Pull request by rgc**
Add support for responseURL where supported.

XMLHttpRequest spec 4.6.1 adds the responseUrl attribute:
https://xhr.spec.whatwg.org/#the-responseurl-attribute

**Browser Support**
Chromium: https://code.google.com/p/chromium/issues/detail?id=377583
Webkit: https://bugs.webkit.org/show_bug.cgi?id=136938
Mozilla: https://bugzilla.mozilla.org/show_bug.cgi?id=998076

A current use case is when the DASH manifest is located at a URL that is redirected via 302 to another location. Currently, the mpdParser will then pull the chunks from the original URI rather than the URl from which the manifest was ultimately downloaded.

**Example - Current Flow**
http://some.host.org/path/manifest.mpd returns 302 to http://other.host.org/path/manifest.mpd.

Subsequent requests for mp4 chunks will be:
http://some.host.org/path/chunk1.mp4
http://some.host.org/path/chunk2.mp4

If the server has been configured to 302 redirect for everything, this will result in each of the above chunks also having a 302 redirect. If the server hasn't been configured, then they will fail.

**Example - Proposed Flow**
http://some.host.org/path/manifest.mpd returns 302 to http://other.host.org/path/manifest.mpd.

Subsequent requests for mp4 chunks will be:
http://other.host.org/path/chunk1.mp4
http://other.host.org/path/chunk2.mp4

As the base_url should be the URI from which the manifest was downloaded.

In browsers without responseUrl support, flow will fallback to current.

### CODE REVIEW

File: lib/util/failover_uri.js [hunk 1]
```diff
@@ -168,6 +168,9 @@ shaka.util.FailoverUri.prototype.createRequest_ =
         // effectively cache every response.
         this.requestPromise_ = null;
         this.request_ = null;
+        if (xhr.responseURL) {
+          this.urls.unshift(new goog.Uri(xhr.responseURL));
```
Comment (joeyparrish) [inline lib/util/failover_uri.js]: What would be better (IMO) than prepending the URL would be to replace the specific URL this request used. (this.urls[i])

**Comment by rgc** [comment]
Agree - replacing is better - will do that.

**Comment by rgc** [comment]
@joeyparrish - Lint/Build passed and I've implemented the replace rather than the unshift. Will read the contrib *before* committing next time  ;)

Thanks!

**Comment by joeyparrish** [comment]
GitHub is saying you have merge conflicts.  Please rebase to the tip of master and manually resolve the merge conflicts.  Thanks!

**Comment by rgc** [comment]
Conflicts have been resolved -- Thanks

**Comment by joeyparrish** [comment]
We are experiencing some flakiness on our buildbot.  Please forgive the noise.

**Comment by rgc** [comment]
I think it is failing b/c the URLs are formatted in the tests as:
//some.host.org/path/dash.mpd
and responseUrl will typically contain the proto and params:
http://some.host.org/path/dash.mpd?m=1212121

So the test is failing as it notices they don't match...

I'll look into it.

**Comment by joeyparrish** [comment]
Yes, that seems to be the case.  Not flake at all.

Here's the log snippet:

```
Chrome 47.0.2526 (Linux 0.0.0) Live live support for segment time template requests MPD update in expected time FAILED
	Expected 'http://storage.googleapis.com/widevine-demo-media/oops-segment-timeline-time/oops-segment-timeline-time-oops_video.mpd?_=1452630896703' to equal '//storage.googleapis.com/widevine-demo-media/oops-segment-timeline-time/oops-segment-timeline-time-oops_video.mpd'.
```

**Comment by rgc** [comment]
I'm pondering the best way to fix.

It looks like we have an expectation that the final MPD location should match the target location:
```javascript
spyOn(window.shaka.dash, 'MpdRequest').and.callFake(function(mpdUrl) {
      expect(mpdUrl.toString()).toEqual(targetMpdUrl);
```

I could strip the proto and params and get it to pass the specific test case, but that's not right.

Ultimately, this expectation isn't going to pass ever in cases where the player was redirected, so I'll need to come up with a better test to fit.

**Comment by joeyparrish** [comment]
Options:

1. Hack XMLHttpRequest to suppress responseURL during the test, then restore it afterward.  delete XMLHttpRequest.prototype['responseURL']
2. Decide it's wrong to expect a specific URL, since this is not relevant to the test.  Remove the targetMpdUrl parameter from waitForMpdRequest().

My preference is for 2.

**Comment by rgc** [comment]
I was testing my solution last night and noticed that by replacing urls[i], we doom future manifest requests. The proper functionality would be for the relative chunks detailed in the manifest to use the 302 location for the relative baseUrl, however future manifest requests should not.

Below is an updated version of flow which encompasses this detail.

**Example - Proposed Flow**
http://some.host.org/path/manifest.mpd --> 302 --> http://other.host.org/path/manifest.mpd.
http://other.host.org/path/chunk1.mp4
http://other.host.org/path/chunk2.mp4
(manifest update)
http://some.host.org/path/manifest.mpd --> 302 --> http://another.host.org/path/manifest.mpd.
http://another.host.org/path/chunk1.mp4
http://another.host.org/path/chunk2.mp4

I've committed a few more changes that fix the flow. I have not altered the failing test, as I believe these new changes may not trigger it.

File: lib/util/failover_uri.js [hunk 2]
```diff
@@ -56,6 +56,10 @@ shaka.util.FailoverUri = function(callback, urls, opt_startByte, opt_endByte) {
 
   /** @private {shaka.util.FailoverUri.NetworkCallback} */
   this.callback_ = callback;
+
+  /** @public {goog.Uri} */
+  this.currentUrl = null;
+
```
Comment (joeyparrish) [inline lib/util/failover_uri.js]: Drop this extra line

File: lib/util/failover_uri.js [hunk 3]
```diff
@@ -56,6 +56,10 @@ shaka.util.FailoverUri = function(callback, urls, opt_startByte, opt_endByte) {
 
   /** @private {shaka.util.FailoverUri.NetworkCallback} */
   this.callback_ = callback;
+
+  /** @public {goog.Uri} */
```
Comment (joeyparrish) [inline lib/util/failover_uri.js]: Just use @type instead of @public

**Comment by joeyparrish** [comment]
One last request:  can you squash these commits together?  It's kind of hard to parse the log with 10 commits, 7 of which have the same message.

**Comment by rgc** [comment]
@joeyparrish - squashed.

Thanks much!

**Comment by joeyparrish** [comment]
@rgc Still getting failures.  I've edited the buildbot message in this thread to just the failures.  (In the mean time, I'm working on getting our buildbot to send the failures only and to format the message correctly.  Should be easier to read if we see failures next time.)

**Comment by rgc** [comment]
@joeyparrish - I was able to run the tests locally, find the issue and fix it.

**Comment by rgc** [comment]
I fixed the spec tests to pass the proper number of arguments to shaka.dash.mpd.parseMpd as well

File: lib/dash/mpd_parser.js [hunk 4]
```diff
@@ -50,8 +51,12 @@ shaka.dash.mpd.parseMpd = function(source, url) {
     return null;
   }
 
+  if (!baseUrl) {
```
Comment (joeyparrish) [inline lib/dash/mpd_parser.js]: You should drop this condition, since it should not be possible after successful compilation.  Otherwise, the PR looks good to me.

**Comment by rgc** [comment]
@joeyparrish - agreed, done and squashed.

**Comment by joeyparrish** [comment]
Looks good!  I'll give the buildbot one more crack at it to verify.  Thanks for taking the time to do revisions and get everything done cleanly.

**Comment by joeyparrish** [comment]
Thanks very much for the contribution.  We will implement something equivalent in Shaka 2 soon to make sure this feature isn't lost in the rewrite.

**Comment by rgc** [comment]
Thanks for walking through the process -- Cheers!
